File: macos/API/Sources/AudioCaptureKit/AudioSession.swift
import Foundation
import AVFoundation
import ScreenCaptureKit

/// Session state enumeration
public enum SessionState: String, Codable {
    case idle = "idle"
    case starting = "starting"
    case active = "active"
    case paused = "paused"
    case stopping = "stopping"
    case stopped = "stopped"
    case error = "error"
}

/// Base class for audio sessions
@available(macOS 13.0, *)
public class BaseAudioSession {
    /// Unique session identifier
    public let id: UUID
    
    /// Session creation time
    public let createdAt: Date
    
    /// Current session state
    private(set) public var state: SessionState = .idle
    
    /// Session statistics
    private(set) public var statistics: SessionStatistics
    
    /// State change observers
    private var stateObservers: [UUID: (SessionState) -> Void] = [:]
    
    /// Error handler
    private var errorHandler: ((Error) -> Void)?
    
    init() {
        self.id = UUID()
        self.createdAt = Date()
        self.statistics = SessionStatistics(
            sessionId: id,
            state: .idle,
            bufferCount: 0,
            duration: 0,
            format: nil
        )
    }
    
    /// Update session state
    func updateState(_ newState: SessionState) {
        let oldState = state
        state = newState
        statistics = SessionStatistics(
            sessionId: id,
            state: newState,
            bufferCount: statistics.bufferCount,
            duration: statistics.duration,
            format: statistics.format
        )
        
        // Notify observers
        for observer in stateObservers.values {
            observer(newState)
        }
    }
    
    /// Add state observer
    public func addStateObserver(_ observer: @escaping (SessionState) -> Void) -> UUID {
        let observerId = UUID()
        stateObservers[observerId] = observer
        return observerId
    }
    
    /// Remove state observer
    public func removeStateObserver(_ observerId: UUID) {
        stateObservers.removeValue(forKey: observerId)
    }
    
    /// Set error handler
    public func setErrorHandler(_ handler: @escaping (Error) -> Void) {
        self.errorHandler = handler
    }
    
    /// Handle error
    func handleError(_ error: Error) {
        updateState(.error)
        errorHandler?(error)
    }
    
    /// Get session statistics
    public func getStatistics() -> SessionStatistics {
        return statistics
    }
}

/// AudioCaptureSession - Manages an audio capture session
@available(macOS 13.0, *)
public actor AudioCaptureSession {
    
    // MARK: - Properties
    
    /// Unique session identifier
    public let id: UUID
    
    /// Session creation time
    public let createdAt: Date
    
    /// Current session state
    private(set) public var state: SessionState = .idle
    
    /// Session statistics
    private(set) public var statistics: SessionStatistics
    
    /// State change observers
    private var stateObservers: [UUID: (SessionState) -> Void] = [:]
    
    /// Error handler
    private var errorHandler: ((Error) -> Void)?
    
    /// Session configuration
    private let configuration: CaptureConfiguration
    
    /// Audio recorder
    private var recorder: StreamingAudioRecorder?
    
    /// Active outputs
    private var outputs: [AudioOutput] = []
    
    /// Stream multiplexer
    private let multiplexer: AudioStreamMultiplexer
    
    /// Format for this session
    private var sessionFormat: AudioFormat?
    
    // MARK: - Initialization
    
    init(configuration: CaptureConfiguration) {
        self.id = UUID()
        self.createdAt = Date()
        self.statistics = SessionStatistics(
            sessionId: id,
            state: .idle,
            bufferCount: 0,
            duration: 0,
            format: nil
        )
        self.configuration = configuration
        self.multiplexer = AudioStreamMultiplexer()
    }
    
    // MARK: - Session Control
    
    /// Start capture session
    public func start() async throws {
        guard state == .idle || state == .stopped else {
            throw AudioCaptureError.invalidState("Session is already active")
        }
        
        await updateState(.starting)
        
        do {
            // Create recorder
            recorder = StreamingAudioRecorder()
            
            // Set up multiplexer as delegate
            recorder?.addStreamDelegate(multiplexer)
            
            // Start recording
            try await recorder?.startStreaming()
            
            // Update format
            sessionFormat = configuration.format ?? AudioFormat.defaultFormat
            
            await updateState(.active)
        } catch {
            await handleError(error)
            throw error
        }
    }
    
    /// Stop capture session
    public func stop() async throws {
        guard state == .active || state == .paused else {
            throw AudioCaptureError.invalidState("Session is not active")
        }
        
        await updateState(.stopping)
        
        // Stop recorder
        await recorder?.stopStreaming()
        
        // Notify outputs
        for output in outputs {
            await output.finish()
        }
        
        // Clear outputs
        outputs.removeAll()
        await multiplexer.removeAllOutputs()
        
        await updateState(.stopped)
    }
    
    /// Pause capture session
    public func pause() async throws {
        guard state == .active else {
            throw AudioCaptureError.invalidState("Session is not active")
        }
        
        await updateState(.paused)
        await multiplexer.setPaused(true)
    }
    
    /// Resume capture session
    public func resume() async throws {
        guard state == .paused else {
            throw AudioCaptureError.invalidState("Session is not paused")
        }
        
        await multiplexer.setPaused(false)
        await updateState(.active)
    }
    
    // MARK: - Output Management
    
    /// Add an output to the session
    public func addOutput(_ output: AudioOutput) async throws {
        guard state == .active || state == .paused else {
            throw AudioCaptureError.invalidState("Session must be active to add outputs")
        }
        
        // Configure output with session format
        if let format = sessionFormat {
            try await output.configure(format: format)
        }
        
        // Add to multiplexer
        await multiplexer.addOutput(output)
        
        // Track output
        outputs.append(output)
    }
    
    /// Remove an output from the session
    public func removeOutput(_ output: AudioOutput) async throws {
        await multiplexer.removeOutput(output)
        outputs.removeAll { $0.id == output.id }
        await output.finish()
    }
    
    /// Get all active outputs
    public func getOutputs() -> [AudioOutput] {
        return outputs
    }
    
    // MARK: - Configuration
    
    /// Get session configuration
    public func getConfiguration() -> CaptureConfiguration {
        return configuration
    }
    
    /// Get session format
    public func getFormat() -> AudioFormat? {
        return sessionFormat
    }
    
    // MARK: - Private Methods
    
    /// Update session state
    func updateState(_ newState: SessionState) {
        let oldState = state
        state = newState
        statistics = SessionStatistics(
            sessionId: id,
            state: newState,
            bufferCount: statistics.bufferCount,
            duration: statistics.duration,
            format: statistics.format
        )
        
        // Notify observers
        for observer in stateObservers.values {
            observer(newState)
        }
    }
    
    /// Handle error
    func handleError(_ error: Error) {
        updateState(.error)
        errorHandler?(error)
    }
    
    /// Get session statistics
    public func getStatistics() -> SessionStatistics {
        return statistics
    }
}

/// AudioPlaybackSession - Manages an audio playback session
@available(macOS 13.0, *)
public actor AudioPlaybackSession {
    
    // MARK: - Properties
    
    /// Unique session identifier
    public let id: UUID
    
    /// Session creation time
    public let createdAt: Date
    
    /// Current session state
    private(set) public var state: SessionState = .idle
    
    /// Session statistics
    private(set) public var statistics: SessionStatistics
    
    /// State change observers
    private var stateObservers: [UUID: (SessionState) -> Void] = [:]
    
    /// Error handler
    private var errorHandler: ((Error) -> Void)?
    
    /// Session configuration
    private let configuration: PlaybackConfiguration
    
    /// Audio player
    private var player: StreamingAudioPlayer?
    
    /// Input source
    private var inputSource: AudioInput?
    
    /// Format for this session
    private var sessionFormat: AudioFormat?
    
    // MARK: - Initialization
    
    init(configuration: PlaybackConfiguration) {
        self.id = UUID()
        self.createdAt = Date()
        self.statistics = SessionStatistics(
            sessionId: id,
            state: .idle,
            bufferCount: 0,
            duration: 0,
            format: nil
        )
        self.configuration = configuration
    }
    
    // MARK: - Session Control
    
    /// Start playback session
    public func start() async throws {
        guard state == .idle || state == .stopped else {
            throw AudioCaptureError.invalidState("Session is already active")
        }
        
        await updateState(.starting)
        
        do {
            // Create player with delay if specified
            player = StreamingAudioPlayer(delay: configuration.delay)
            
            // Set volume
            player?.volume = configuration.volume
            
            // Start playback
            try player?.startPlayback()
            
            // Update format
            sessionFormat = configuration.format ?? AudioFormat.defaultFormat
            
            await updateState(.active)
        } catch {
            await handleError(error)
            throw error
        }
    }
    
    /// Stop playback session
    public func stop() async throws {
        guard state == .active || state == .paused else {
            throw AudioCaptureError.invalidState("Session is not active")
        }
        
        await updateState(.stopping)
        
        // Stop player
        player?.stopPlayback()
        
        // Disconnect input
        if let input = inputSource {
            await input.disconnect()
        }
        
        await updateState(.stopped)
    }
    
    /// Pause playback session
    public func pause() async throws {
        guard state == .active else {
            throw AudioCaptureError.invalidState("Session is not active")
        }
        
        // TODO: Implement pause in StreamingAudioPlayer
        await updateState(.paused)
    }
    
    /// Resume playback session
    public func resume() async throws {
        guard state == .paused else {
            throw AudioCaptureError.invalidState("Session is not paused")
        }
        
        // TODO: Implement resume in StreamingAudioPlayer
        await updateState(.active)
    }
    
    // MARK: - Input Management
    
    /// Set input source for playback
    public func setInput(_ input: AudioInput) async throws {
        guard state == .active || state == .paused else {
            throw AudioCaptureError.invalidState("Session must be active to set input")
        }
        
        // Disconnect previous input
        if let currentInput = inputSource {
            await currentInput.disconnect()
        }
        
        // Connect new input
        self.inputSource = input
        
        if let player = player {
            try await input.connect(to: player)
        }
    }
    
    /// Get current input source
    public func getInput() -> AudioInput? {
        return inputSource
    }
    
    // MARK: - Playback Control
    
    /// Set playback volume (0.0 to 1.0)
    public func setVolume(_ volume: Float) {
        player?.volume = max(0.0, min(1.0, volume))
    }
    
    /// Get current volume
    public func getVolume() -> Float {
        return player?.volume ?? configuration.volume
    }
    
    // MARK: - Buffer Management
    
    /// Schedule a buffer for playback
    public func scheduleBuffer(_ buffer: AVAudioPCMBuffer) {
        player?.scheduleBuffer(buffer)
    }
    
    // MARK: - Configuration
    
    /// Get session configuration
    public func getConfiguration() -> PlaybackConfiguration {
        return configuration
    }
    
    /// Get session format
    public func getFormat() -> AudioFormat? {
        return sessionFormat
    }
    
    // MARK: - Private Methods
    
    /// Update session state
    func updateState(_ newState: SessionState) {
        let oldState = state
        state = newState
        statistics = SessionStatistics(
            sessionId: id,
            state: newState,
            bufferCount: statistics.bufferCount,
            duration: statistics.duration,
            format: statistics.format
        )
        
        // Notify observers
        for observer in stateObservers.values {
            observer(newState)
        }
    }
    
    /// Handle error
    func handleError(_ error: Error) {
        updateState(.error)
        errorHandler?(error)
    }
    
    /// Get session statistics
    public func getStatistics() -> SessionStatistics {
        return statistics
    }
}

/// AudioStreamMultiplexer - Distributes audio to multiple outputs
@available(macOS 13.0, *)
actor AudioStreamMultiplexer: AudioStreamDelegate {
    
    // MARK: - Properties
    
    /// Active outputs
    private var outputs: [AudioOutput] = []
    
    /// Processing queue
    private let processingQueue = DispatchQueue(label: "com.audiocapture.multiplexer", qos: .userInitiated)
    
    /// Paused state
    private var isPaused = false
    
    /// Buffer count for statistics
    private var bufferCount = 0
    
    // MARK: - Output Management
    
    func addOutput(_ output: AudioOutput) {
        outputs.append(output)
    }
    
    func removeOutput(_ output: AudioOutput) {
        outputs.removeAll { $0.id == output.id }
    }
    
    func removeAllOutputs() {
        outputs.removeAll()
    }
    
    func setPaused(_ paused: Bool) {
        isPaused = paused
    }
    
    // MARK: - AudioStreamDelegate
    
    nonisolated func audioStreamer(_ streamer: StreamingAudioRecorder, didReceive buffer: AVAudioPCMBuffer) {
        Task {
            await processBuffer(buffer)
        }
    }
    
    nonisolated func audioStreamer(_ streamer: StreamingAudioRecorder, didEncounterError error: Error) {
        Task {
            await handleError(error)
        }
    }
    
    nonisolated func audioStreamerDidFinish(_ streamer: StreamingAudioRecorder) {
        Task {
            await finish()
        }
    }
    
    // MARK: - Private Methods
    
    private func processBuffer(_ buffer: AVAudioPCMBuffer) async {
        guard !isPaused else { return }
        
        bufferCount += 1
        
        // Create audio buffer wrapper
        let audioBuffer = AudioBuffer(pcmBuffer: buffer)
        
        // Distribute to all outputs
        await withTaskGroup(of: Void.self) { group in
            for output in outputs {
                group.addTask {
                    do {
                        try await output.process(audioBuffer)
                    } catch {
                        print("Output \(output.id) failed to process buffer: \(error)")
                    }
                }
            }
        }
    }
    
    private func handleError(_ error: Error) async {
        // Notify all outputs of error
        for output in outputs {
            await output.handleError(error)
        }
    }
    
    private func finish() async {
        // Notify all outputs of finish
        for output in outputs {
            await output.finish()
        }
    }
}

/// AudioInput protocol for playback sources
public protocol AudioInput {
    /// Unique identifier
    var id: UUID { get }
    
    /// Connect to a player
    func connect(to player: StreamingAudioPlayer) async throws
    
    /// Disconnect from player
    func disconnect() async
}

/// Capture session as audio input
@available(macOS 13.0, *)
extension AudioCaptureSession: AudioInput {
    public func connect(to player: StreamingAudioPlayer) async throws {
        // Create bridge output that forwards to player
        let bridgeOutput = BridgeOutput(player: player)
        try await addOutput(bridgeOutput)
    }
    
    public func disconnect() async {
        // Remove bridge outputs
        let bridgeOutputs = outputs.filter { $0 is BridgeOutput }
        for output in bridgeOutputs {
            try? await removeOutput(output)
        }
    }
}

/// Bridge output for connecting capture to playback
@available(macOS 13.0, *)
class BridgeOutput: AudioOutput {
    let id = UUID()
    private let player: StreamingAudioPlayer
    
    init(player: StreamingAudioPlayer) {
        self.player = player
    }
    
    func configure(format: AudioFormat) async throws {
        // Player handles format internally
    }
    
    func process(_ buffer: AudioBuffer) async throws {
        player.scheduleBuffer(buffer.pcmBuffer)
    }
    
    func handleError(_ error: Error) async {
        print("Bridge output error: \(error)")
    }
    
    func finish() async {
        // Nothing to cleanup
    }
}

File: windows/API/AudioSession.py
"""
AudioSession - Session management for audio capture and playback

This module provides session management for audio operations, including
state tracking, output management, and configuration handling.
"""

import asyncio
import uuid
from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum, auto
from typing import List, Dict, Optional, Callable, Any
import numpy as np

from .AudioFormat import AudioFormat, AudioBuffer
from .AudioOutput import AudioOutput
from .AudioDevice import AudioDevice
from .StreamingAudioRecorder import StreamingAudioRecorder, AudioStreamDelegate
from .StreamingAudioPlayer import StreamingAudioPlayer
from .AudioError import InvalidStateError, SessionNotFoundError


class SessionState(Enum):
    """Session state enumeration"""
    IDLE = "idle"
    STARTING = "starting"
    ACTIVE = "active"
    PAUSED = "paused"
    STOPPING = "stopping"
    STOPPED = "stopped"
    ERROR = "error"


@dataclass
class SessionStatistics:
    """Session statistics"""
    session_id: uuid.UUID
    state: SessionState
    buffer_count: int
    duration: float
    format: Optional[AudioFormat]
    created_at: datetime = field(default_factory=datetime.now)
    
    @property
    def uptime(self) -> float:
        """Get session uptime in seconds"""
        return (datetime.now() - self.created_at).total_seconds()


@dataclass
class CaptureConfiguration:
    """Audio capture configuration"""
    device: Optional[AudioDevice] = None
    format: Optional[AudioFormat] = None
    buffer_size: int = 1024
    
    def __post_init__(self):
        if self.format is None:
            self.format = AudioFormat.default_format()


@dataclass
class PlaybackConfiguration:
    """Audio playback configuration"""
    device: Optional[AudioDevice] = None
    format: Optional[AudioFormat] = None
    volume: float = 1.0
    delay: float = 0.0
    buffer_size: int = 1024
    
    def __post_init__(self):
        if self.format is None:
            self.format = AudioFormat.default_format()
        self.volume = max(0.0, min(1.0, self.volume))
        self.delay = max(0.0, self.delay)


class AudioStreamMultiplexer(AudioStreamDelegate):
    """Distributes audio to multiple outputs"""
    
    def __init__(self):
        """Initialize audio stream multiplexer"""
        self._outputs: List[AudioOutput] = []
        self._is_paused = False
        self._buffer_count = 0
        self._lock = asyncio.Lock()
    
    async def add_output(self, output: AudioOutput) -> None:
        """Add an output"""
        async with self._lock:
            if output not in self._outputs:
                self._outputs.append(output)
    
    async def remove_output(self, output: AudioOutput) -> None:
        """Remove an output"""
        async with self._lock:
            if output in self._outputs:
                self._outputs.remove(output)
    
    async def remove_all_outputs(self) -> None:
        """Remove all outputs"""
        async with self._lock:
            self._outputs.clear()
    
    async def set_paused(self, paused: bool) -> None:
        """Set paused state"""
        self._is_paused = paused
    
    async def audio_streamer_did_receive(self, streamer: StreamingAudioRecorder, buffer: AudioBuffer):
        """Handle received audio buffer"""
        if self._is_paused:
            return
        
        self._buffer_count += 1
        
        # Distribute to all outputs
        async with self._lock:
            outputs = list(self._outputs)
        
        # Process outputs concurrently
        tasks = []
        for output in outputs:
            tasks.append(self._process_output(output, buffer))
        
        if tasks:
            await asyncio.gather(*tasks, return_exceptions=True)
    
    async def _process_output(self, output: AudioOutput, buffer: AudioBuffer):
        """Process buffer for a single output"""
        try:
            await output.process(buffer)
        except Exception as e:
            print(f"Output {output.id} failed to process buffer: {e}")
    
    async def audio_streamer_did_encounter_error(self, streamer: StreamingAudioRecorder, error: Exception):
        """Handle error from streamer"""
        async with self._lock:
            outputs = list(self._outputs)
        
        for output in outputs:
            try:
                await output.handle_error(error)
            except Exception:
                pass
    
    async def audio_streamer_did_finish(self, streamer: StreamingAudioRecorder):
        """Handle streamer finish"""
        async with self._lock:
            outputs = list(self._outputs)
        
        for output in outputs:
            try:
                await output.finish()
            except Exception:
                pass


class BaseAudioSession:
    """Base class for audio sessions"""
    
    def __init__(self):
        """Initialize base audio session"""
        self.id = uuid.uuid4()
        self.created_at = datetime.now()
        self._state = SessionState.IDLE
        self._statistics = SessionStatistics(
            session_id=self.id,
            state=SessionState.IDLE,
            buffer_count=0,
            duration=0.0,
            format=None
        )
        self._state_observers: Dict[uuid.UUID, Callable[[SessionState], None]] = {}
        self._error_handler: Optional[Callable[[Exception], None]] = None
        self._state_lock = asyncio.Lock()
    
    async def update_state(self, new_state: SessionState) -> None:
        """Update session state"""
        async with self._state_lock:
            old_state = self._state
            self._state = new_state
            self._statistics.state = new_state
            
            # Notify observers
            observers = list(self._state_observers.values())
        
        # Notify outside of lock
        for observer in observers:
            try:
                observer(new_state)
            except Exception as e:
                print(f"State observer error: {e}")
    
    def add_state_observer(self, observer: Callable[[SessionState], None]) -> uuid.UUID:
        """
        Add state observer.
        
        Args:
            observer: Callback function for state changes
            
        Returns:
            Observer ID for removal
        """
        observer_id = uuid.uuid4()
        self._state_observers[observer_id] = observer
        return observer_id
    
    def remove_state_observer(self, observer_id: uuid.UUID) -> None:
        """Remove state observer"""
        self._state_observers.pop(observer_id, None)
    
    def set_error_handler(self, handler: Callable[[Exception], None]) -> None:
        """Set error handler"""
        self._error_handler = handler
    
    async def handle_error(self, error: Exception) -> None:
        """Handle error"""
        await self.update_state(SessionState.ERROR)
        if self._error_handler:
            try:
                self._error_handler(error)
            except Exception:
                pass
    
    def get_statistics(self) -> SessionStatistics:
        """Get session statistics"""
        return self._statistics
    
    @property
    def state(self) -> SessionState:
        """Get current state"""
        return self._state


class AudioCaptureSession(BaseAudioSession):
    """Manages an audio capture session"""
    
    def __init__(self, configuration: CaptureConfiguration):
        """
        Initialize audio capture session.
        
        Args:
            configuration: Capture configuration
        """
        super().__init__()
        self._configuration = configuration
        self._recorder: Optional[StreamingAudioRecorder] = None
        self._outputs: List[AudioOutput] = []
        self._multiplexer = AudioStreamMultiplexer()
        self._session_format = configuration.format
    
    async def start(self) -> None:
        """Start capture session"""
        if self._state not in (SessionState.IDLE, SessionState.STOPPED):
            raise InvalidStateError("Session is already active")
        
        await self.update_state(SessionState.STARTING)
        
        try:
            # Create recorder
            self._recorder = StreamingAudioRecorder(
                sample_rate=int(self._session_format.sample_rate),
                channels=self._session_format.channel_count,
                blocksize=self._configuration.buffer_size,
                device=self._configuration.device
            )
            
            # Set up multiplexer as delegate
            self._recorder.add_delegate(self._multiplexer)
            
            # Start recording
            await self._recorder.start_streaming()
            
            # Update statistics
            self._statistics.format = self._session_format
            
            await self.update_state(SessionState.ACTIVE)
            
        except Exception as e:
            await self.handle_error(e)
            raise
    
    async def stop(self) -> None:
        """Stop capture session"""
        if self._state not in (SessionState.ACTIVE, SessionState.PAUSED):
            raise InvalidStateError("Session is not active")
        
        await self.update_state(SessionState.STOPPING)
        
        # Stop recorder
        if self._recorder:
            await self._recorder.stop_streaming()
        
        # Notify outputs
        for output in self._outputs:
            try:
                await output.finish()
            except Exception:
                pass
        
        # Clear outputs
        self._outputs.clear()
        await self._multiplexer.remove_all_outputs()
        
        await self.update_state(SessionState.STOPPED)
    
    async def pause(self) -> None:
        """Pause capture session"""
        if self._state != SessionState.ACTIVE:
            raise InvalidStateError("Session is not active")
        
        await self.update_state(SessionState.PAUSED)
        await self._multiplexer.set_paused(True)
    
    async def resume(self) -> None:
        """Resume capture session"""
        if self._state != SessionState.PAUSED:
            raise InvalidStateError("Session is not paused")
        
        await self._multiplexer.set_paused(False)
        await self.update_state(SessionState.ACTIVE)
    
    async def add_output(self, output: AudioOutput) -> None:
        """
        Add an output to the session.
        
        Args:
            output: Audio output to add
        """
        if self._state not in (SessionState.ACTIVE, SessionState.PAUSED):
            raise InvalidStateError("Session must be active to add outputs")
        
        # Configure output with session format
        if self._session_format:
            await output.configure(self._session_format)
        
        # Add to multiplexer
        await self._multiplexer.add_output(output)
        
        # Track output
        self._outputs.append(output)
    
    async def remove_output(self, output: AudioOutput) -> None:
        """Remove an output from the session"""
        await self._multiplexer.remove_output(output)
        if output in self._outputs:
            self._outputs.remove(output)
        await output.finish()
    
    def get_outputs(self) -> List[AudioOutput]:
        """Get all active outputs"""
        return list(self._outputs)
    
    def get_configuration(self) -> CaptureConfiguration:
        """Get session configuration"""
        return self._configuration
    
    def get_format(self) -> Optional[AudioFormat]:
        """Get session format"""
        return self._session_format


class AudioPlaybackSession(BaseAudioSession):
    """Manages an audio playback session"""
    
    def __init__(self, configuration: PlaybackConfiguration):
        """
        Initialize audio playback session.
        
        Args:
            configuration: Playback configuration
        """
        super().__init__()
        self._configuration = configuration
        self._player: Optional[StreamingAudioPlayer] = None
        self._input_source = None
        self._session_format = configuration.format
    
    async def start(self) -> None:
        """Start playback session"""
        if self._state not in (SessionState.IDLE, SessionState.STOPPED):
            raise InvalidStateError("Session is already active")
        
        await self.update_state(SessionState.STARTING)
        
        try:
            # Create player
            device_index = self._configuration.device.device_index if self._configuration.device else None
            self._player = StreamingAudioPlayer(
                sample_rate=int(self._session_format.sample_rate),
                channels=self._session_format.channel_count,
                device_index=device_index,
                delay=self._configuration.delay,
                blocksize=self._configuration.buffer_size
            )
            
            # Set volume
            self._player.set_volume(self._configuration.volume)
            
            # Start playback
            await self._player.start_playback()
            
            # Update statistics
            self._statistics.format = self._session_format
            
            await self.update_state(SessionState.ACTIVE)
            
        except Exception as e:
            await self.handle_error(e)
            raise
    
    async def stop(self) -> None:
        """Stop playback session"""
        if self._state not in (SessionState.ACTIVE, SessionState.PAUSED):
            raise InvalidStateError("Session is not active")
        
        await self.update_state(SessionState.STOPPING)
        
        # Stop player
        if self._player:
            await self._player.stop_playback()
        
        # Disconnect input
        if self._input_source:
            await self._input_source.disconnect()
        
        await self.update_state(SessionState.STOPPED)
    
    async def pause(self) -> None:
        """Pause playback session"""
        if self._state != SessionState.ACTIVE:
            raise InvalidStateError("Session is not active")
        
        # TODO: Implement pause in StreamingAudioPlayer
        await self.update_state(SessionState.PAUSED)
    
    async def resume(self) -> None:
        """Resume playback session"""
        if self._state != SessionState.PAUSED:
            raise InvalidStateError("Session is not paused")
        
        # TODO: Implement resume in StreamingAudioPlayer
        await self.update_state(SessionState.ACTIVE)
    
    async def schedule_buffer(self, buffer: AudioBuffer) -> None:
        """Schedule a buffer for playback"""
        if not self._player:
            raise InvalidStateError("Player not initialized")
        
        await self._player.schedule_buffer(buffer.data)
        self._statistics.buffer_count += 1
    
    def set_volume(self, volume: float) -> None:
        """Set playback volume (0.0 to 1.0)"""
        if self._player:
            self._player.set_volume(volume)
    
    def get_volume(self) -> float:
        """Get current volume"""
        if self._player:
            return self._player.get_volume()
        return self._configuration.volume
    
    def get_configuration(self) -> PlaybackConfiguration:
        """Get session configuration"""
        return self._configuration
    
    def get_format(self) -> Optional[AudioFormat]:
        """Get session format"""
        return self._session_format